### DISCUSSION TITLE
What is the proper way to pass something into slash command handler?

### DISCUSSION THREAD

**Discussion by telephrag** [bwmarrin/discordgo#1113] [Q&A] [answered]
I'm interacting with mongodb inside handlers and need pointer to collection inside database.

**Comment by FedorLap2006** [comment, marked as answer]
I think [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is what you need.
In Go you can implement it using structs and methods. Like that:
```go
type MyAppOrModule struct {
     Client *Something
}

func (m *MyAppOrModule) HandleSlash(s *discordgo.Session, i *discordgo.InteractionCreate) {
     m.Client.DoSomething(...)
}
```

**Reply by telephrag** [reply]
```go

var HandlerToCmd = map[string]func(s *discordgo.Session, i *discordgo.InteractionCreate){
    "command": nil,
    // other commands...
}

type MongoSession struct {
    Client     *mongo.Client
    Collection *mongo.Collection
}

func New(args) *MongoSession {
    ms := &MongoSession{
        // ...
    }

    HandlerToCmd["command"] = (ms).Command
    // assigning other methods to map...
}

func DoSmtWithDb(i *discordgo.InteractionCreate, ms* MongoSession) {
    // ...
}

func (ms *MongoSession) Command(s *discordgo.Session, i *discordgo.InteractionCreate) {
    // do stuff...
    DoSmtWithDb(i, ms)
}
```

This is what I ended up with. It works but can it be made simpler? It used to be just global variable that pointed to `MongoSession` object.

**Reply by FedorLap2006** [reply]
I would recommend renaming it to `Module` or `Instance` / `<BotName>`, that's the general approach for that.
Additionally, I think would be better to move all the logic inside the handler.
But if you're going with "Clean Architecture" approach, you can make services and repositories which will take care of the storage and business logic, leaving handlers only to work with input and output.

**Reply by telephrag** [reply]
Thank you.
